src/analyses/base.ml

```diff
@@ -1809,88 +1824,89 @@ struct
       | Some exp ->
         let t_override = match Cilfacade.fundec_return_type fundec with
           | TVoid _ -> M.warn ~category:M.Category.Program "Returning a value from a void function"; assert false
           | ret -> ret
         in
-        let rv = eval_rv (Analyses.ask_of_ctx ctx) ctx.global ctx.local exp in
-        let st' = set ~ctx ~t_override (Analyses.ask_of_ctx ctx) ctx.global nst (return_var ()) t_override rv in
-        match ThreadId.get_current (Analyses.ask_of_ctx ctx) with
-        | `Lifted tid when ThreadReturn.is_current (Analyses.ask_of_ctx ctx) ->
+        let rv = eval_rv ~ctx ctx.local exp in
+        let st' = set ~ctx ~t_override nst (return_var ()) t_override rv in
+        match ThreadId.get_current ask with
+        | `Lifted tid when ThreadReturn.is_current ask ->
           (* Evaluate exp and cast the resulting value to the void-pointer-type.
               Casting to the right type here avoids precision loss on joins. *)
           let rv = VD.cast ~torg:(Cilfacade.typeOf exp) Cil.voidPtrType rv in
           ctx.sideg (V.thread tid) (G.create_thread rv);
-          Priv.thread_return (Analyses.ask_of_ctx ctx) (priv_getg ctx.global) (priv_sideg ctx.sideg) tid st'
+          Priv.thread_return ask (priv_getg ctx.global) (priv_sideg ctx.sideg) tid st'
         | _ -> st'
 
   let vdecl ctx (v:varinfo) =
     if not (Cil.isArrayType v.vtype) then
       ctx.local
     else
-      let lval = eval_lv (Analyses.ask_of_ctx ctx) ctx.global ctx.local (Var v, NoOffset) in
-      let current_value = eval_rv (Analyses.ask_of_ctx ctx) ctx.global ctx.local (Lval (Var v, NoOffset)) in
-      let new_value = VD.update_array_lengths (eval_rv (Analyses.ask_of_ctx ctx) ctx.global ctx.local) current_value v.vtype in
-      set ~ctx (Analyses.ask_of_ctx ctx) ctx.global ctx.local lval v.vtype new_value
+      let lval = eval_lv ~ctx ctx.local (Var v, NoOffset) in
+      let current_value = eval_rv ~ctx ctx.local (Lval (Var v, NoOffset)) in
+      let new_value = VD.update_array_lengths (eval_rv ~ctx ctx.local) current_value v.vtype in
+      set ~ctx ctx.local lval v.vtype new_value
 
   (**************************************************************************
    * Function calls
    **************************************************************************)
 
   (** From a list of expressions, collect a list of addresses that they might point to, or contain pointers to. *)
-  let collect_funargs ask ?(warn=false) (gs:glob_fun) (st:store) (exps: exp list) =
+  let collect_funargs ~ctx ?(warn=false) (st:store) (exps: exp list) =
     let do_exp e =
-      let immediately_reachable = reachable_from_value ask gs st (eval_rv ask gs st e) (Cilfacade.typeOf e) (CilType.Exp.show e) in
-      reachable_vars ask [immediately_reachable] gs st
+      let immediately_reachable = reachable_from_value ~ctx (eval_rv ~ctx st e) (Cilfacade.typeOf e) (CilType.Exp.show e) in
+      reachable_vars ~ctx [immediately_reachable]
```

Comment:
I'm not sure I quite get why is this guaranteed to be the same? To me, it looks like we are evaluating some things on `st` and others on `ctx.local`?

Reply:
This PR currently leaves `st` as separate from `ctx.local` because of this, but there's a broader conceptual problem with it even before this PR.
Namely, even if a different `st` is passed and used for some local lookups, `ask` isn't (and cannot be) modified, so `EvalInt` query would still run on the original local state. So having a separate local state is overall questionable and should maybe be addressed separately in the future somehow.